compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -1468,7 +1527,14 @@ public void visit(BLangAssignment assignNode) {
         setTypeOfVarRefInAssignment(varRef);
         expType = varRef.type;
 
-        typeChecker.checkExpr(assignNode.expr, this.env, expType);
+        BType typeChecked = typeChecker.checkExpr(assignNode.expr, this.env, expType);
+
+        if (typeChecked != symTable.semanticError && varRef.getKind() == NodeKind.SIMPLE_VARIABLE_REF) {
+            BLangSimpleVarRef variableRef = (BLangSimpleVarRef) varRef;
+            if (variableRef.lhsVar && variableRef.type.tsymbol == null && variableRef.type.tag == TypeTags.TYPEDESC) {
+                dlog.error(variableRef.pos, DiagnosticErrorCode.INVALID_VARIABLE_REFERENCE);
+            }
+        }
 
```

Comment:
```ballerina
type x record {|
|};
public function main() {
    x = 5;
}
```
I think this is not the correct fix. For the above code, It gives an error as `incompatible types: expected 'typedesc<x>', found 'int'` (Here, the issue is `x`(type definition) taken as the symbol of the variable reference `x`(x = 5). We need to fix it(#30311)). @MaryamZi

Reply:
Yeah I think we need to say we can't assign to `x` here.